hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java

```diff
@@ -856,19 +860,34 @@ private boolean moveContainer(DatanodeDetails source,
                     " {} failed: {}",
                 moveSelection.getContainerID(), source.getUuidString(),
                 moveSelection.getTargetNode().getUuidString(), result);
+            // add source back if move failed due to container related errors
+            if (result == MoveManager.MoveResult.REPLICATION_FAIL_NOT_EXIST_IN_SOURCE ||
+                result == MoveManager.MoveResult.REPLICATION_FAIL_EXIST_IN_TARGET ||
+                result == MoveManager.MoveResult.REPLICATION_FAIL_CONTAINER_NOT_CLOSED ||
+                result == MoveManager.MoveResult.REPLICATION_FAIL_INFLIGHT_DELETION ||
+                result == MoveManager.MoveResult.REPLICATION_FAIL_INFLIGHT_REPLICATION) {
+              findSourceStrategy.addBackSourceDataNode(source);
+            }
```

Comment:
Since the `whenComplete` is process after the move work, which is async, I'm not sure if this will help with the moveSelection, the selection part might have finished already.

Reply:
In my understanding, in case the move is completed within one iteration of the containerBalancer, then in the next run of the infinite loop wouldn't this change would help by adding the source back? What I mean is:
```
ContainerBalancer # doIteration() {
  ...  //this logic runs for each iteration
  while(true) {
    ...
    // in one run of this loop, the source is selected and move is scheduled
    // in some consecutive run, the previous move is completed with false result, 
    //         so following current PR changes, the source is added back.
    // so now it can be considered for source selection next time
    ...
  }
  ...
}
```
If we see in doIteration -> processMoveSelection -> updateTargetsAndSelectionCriteria -> the same is done if move was successful. 
This change simply does the same if move is failing too.

Reply:
@Tejaskriya After scheduling the move here:
```
      if (replicationManager.getConfig().isLegacyEnabled()) {
        future = replicationManager
            .move(containerID, source, moveSelection.getTargetNode());
      } else {
        future = moveManager.move(containerID, source,
            moveSelection.getTargetNode());
      }
```
this method will return true here if the move hasn't failed/completed yet:
```
      moveSelectionToFutureMap.put(moveSelection, future);
      return true;

```

Then, `updateTargetsAndSelectionCriteria` will eventually call `findSourceStrategy.increaseSizeLeaving(source, size);`, which will add the source datanode back so it can be considered in further move selections. So, we don't need to add the source back in the asynchronous `whenComplete` handling of the `future`.

Reply:
We need to handle the case where the scheduled move completes with a `MoveResult` other than `COMPLETED`, before the if...else code of `if (future.isDone())` gets executed. In this case, this line `return result == MoveManager.MoveResult.COMPLETED;` will evaluate to false and the source datanode will not get added back. However, instead of handling this case in the whenComplete block, I suggest handling it this way:
```
    if (future.isDone()) {
      if (future.isCompletedExceptionally()) {
        return false;
      } else {
        MoveManager.MoveResult result = future.join();
        moveSelectionToFutureMap.put(moveSelection, future);
        if (result == MoveManager.MoveResult.REPLICATION_FAIL_NOT_EXIST_IN_SOURCE ||
            result == MoveManager.MoveResult.REPLICATION_FAIL_EXIST_IN_TARGET ||
            result == MoveManager.MoveResult.REPLICATION_FAIL_CONTAINER_NOT_CLOSED ||
            result == MoveManager.MoveResult.REPLICATION_FAIL_INFLIGHT_DELETION ||
            result == MoveManager.MoveResult.REPLICATION_FAIL_INFLIGHT_REPLICATION) {
          findSourceStrategy.addBackSourceDataNode(source);
        }
        return result == MoveManager.MoveResult.COMPLETED;
      }
    } else {
      moveSelectionToFutureMap.put(moveSelection, future);
      return true;
    }
```
 
This will achieve the same result and I think is a better way mainly because of thread safety issues. Any code in the `whenComplete` block can be called from a thread other than the balancing thread. This means we need to care about thread safety problems, such as concurrent accesses to the priority queue that `findSourceStrategy.addBackSourceDataNode(source)` is changing. We're better off avoiding this to prevent multithreading issues.

Reply:
I have followed this method to handle the checks. Thank you for the suggestion. Could you please review the recent patch?